File: problems/CSES/Sorting_and_Searching/subarray_divisibility/subarray_divisibility.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n = int(input())
    arr = map(int, input().split())

    cnt = [0] * n
    cnt[0] = 1

    pref = 0
    res = 0

    for x in arr:
        pref = (pref + x) % n

        res += cnt[pref]
        cnt[pref] += 1

    print(res)

main()

File: problems/CSES/Sorting_and_Searching/subarray_divisibility/subarray_divisibility.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<ll> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];

    vector<ll> count(n, 0);
    count[0] = 1;

    ll pref = 0;
    ll res = 0;

    for (ll x : arr) {
        pref = (pref + x) % n;

        if (pref < 0) pref += n;

        res += count[pref];
        count[pref]++;
    }

    cout << res << "\n";

    return 0;
}